modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -0,0 +1,68 @@
+package org.tessellation.infrastructure.snapshot
+
+import cats.Applicative
+import cats.data.NonEmptyList
+import cats.effect.{Async, Ref}
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.order._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.dag.snapshot.{GlobalSnapshot, SnapshotOrdinal, StateChannelSnapshotWrapper}
+import org.tessellation.domain.snapshot.GlobalSnapshotStorage
+import org.tessellation.ext.crypto._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address._
+import org.tessellation.security.hash.Hash
+
+object GlobalSnapshotStorage {
+
+  def make[F[_]: Async: KryoSerializer](genesis: GlobalSnapshot): F[GlobalSnapshotStorage[F]] =
+    Ref[F].of[NonEmptyList[GlobalSnapshot]](NonEmptyList.of(genesis)).map(make(_))
+
+  def make[F[_]: Async: KryoSerializer](
+    snapshots: Ref[F, NonEmptyList[GlobalSnapshot]]
+  ): GlobalSnapshotStorage[F] = new GlobalSnapshotStorage[F] {
+
+    def save(snapshot: GlobalSnapshot): F[Unit] =
+      for {
+        _ <- snapshots.modify { nel =>
+          val lastSnapshot = nel.head
+          lastSnapshot.hash match {
+            case Left(error) => (nel, error.raiseError[F, Unit])
+            case Right(lastSnapshotHash) =>
+              val expectedLink = (lastSnapshotHash, SnapshotOrdinal.next.next(lastSnapshot.ordinal))
+              val actualLink = (snapshot.lastSnapshotHash, snapshot.ordinal)
+              if (expectedLink === actualLink) {
+                (snapshot :: nel, Applicative[F].unit)
+              } else {
+                (nel, InvalidGlobalSnapshotChain(expectedLink, actualLink).raiseError[F, Unit])
+              }
+          }
+        }.flatten
+      } yield ()
+
+    def get(ordinal: SnapshotOrdinal): F[Option[GlobalSnapshot]] =
+      snapshots.get.map(_.find(_.ordinal === ordinal))
+
+    def getLast: F[GlobalSnapshot] = snapshots.get.map(_.head)
+
+    def getStateChannelSnapshotBelowOrdinal(
```

Comment:
Implementation looks like it is more like 'UntilOrdinal' / `AtOrBelowOrdinal` or something. I mean it may also be the specified ordinal. It's not always gonna be below the ordinal.

Reply:
You're right. `until` sounds good.